com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderProgramBuilder.java

```diff
@@ -151,27 +153,53 @@ static public String compileGLSL(String shaderSource, ES2ToES3Converter.ShaderTy
             ES2ToES3Converter.Result es3Result = ES2ToES3Converter.transform(source, shaderType, gles ? "es" : "", version, false);
             source = es3Result.output;
         }
+
         return source;
     }
 
-    private ShaderDesc.Shader.Builder buildGLSL(ByteArrayInputStream is, ES2ToES3Converter.ShaderType shaderType, ShaderDesc.Language shaderLanguage,
-                                IResource resource, String resourceOutput, boolean isDebug)  throws IOException, CompileExceptionError {
-        ShaderDesc.Shader.Builder builder = ShaderDesc.Shader.newBuilder();
-        builder.setLanguage(shaderLanguage);
-
+    static private String byteArrayInputStreamToString(ByteArrayInputStream is) {
         int n = is.available();
         byte[] bytes = new byte[n];
         is.read(bytes, 0, n);
-        String source = new String(bytes, StandardCharsets.UTF_8);
+        return new String(bytes, StandardCharsets.UTF_8);
+    }
+
+    private ShaderDesc.Shader.Builder[] buildGLSL(String source, ES2ToES3Converter.ShaderType shaderType, ShaderDesc.Language shaderLanguage,
+                                IResource resource, String resourceOutput, boolean isDebug)  throws IOException, CompileExceptionError {
+
+        ArrayList<ShaderDesc.Shader.Builder> variantBuilders = new ArrayList<ShaderDesc.Shader.Builder>();
 
+        ///////////////////////////////////////////////////////
+        // Build the base variant with just the incoming source
+        ///////////////////////////////////////////////////////
+        ShaderDesc.Shader.Builder baseBuilder = ShaderDesc.Shader.newBuilder();
         String transformedSource = compileGLSL(source, shaderType, shaderLanguage, resourceOutput, isDebug);
+        baseBuilder.setLanguage(shaderLanguage);
+        baseBuilder.setSource(ByteString.copyFrom(transformedSource, "UTF-8"));
+        variantBuilders.add(baseBuilder);
+
+        boolean gles2Standard = shaderLanguage == ShaderDesc.Language.LANGUAGE_GLSL_SM120 ||
+                                shaderLanguage == ShaderDesc.Language.LANGUAGE_GLES_SM100;
+
+        if (gles2Standard) {
+            ///////////////////////////////////////////////////////
+            // Texture array variant
+            ///////////////////////////////////////////////////////
+            String sourceVariantTextureArray = ES2Variants.variantTextureArrayFallback(source);
+            if (sourceVariantTextureArray != null) {
+                ShaderDesc.Shader.Builder builder = ShaderDesc.Shader.newBuilder();
+                sourceVariantTextureArray = compileGLSL(sourceVariantTextureArray, shaderType, shaderLanguage, resourceOutput, isDebug);
+                builder.setLanguage(shaderLanguage);
+                builder.setSource(ByteString.copyFrom(sourceVariantTextureArray, "UTF-8"));
+                builder.setVariantTextureArray(true);
+                variantBuilders.add(builder);
+            }
+        }
```

Comment:
For ES2 we need to also build the fallback shader that should be used on that platform. That process basically inserts the custom page sampling function and replaces all the sampler2DArray calls with our own function (dm_sampler_2d_array or whatever it's called).